### PR TITLE
feat(automated-checks): add listeners for assessment actions in AssessmentCardSelectionStore

### PR DESCRIPTION
**Pull request by lisli1**
#### Details

This PR adds listeners for various assessment actions (scanCompleted, resetData, resetAllAssessmentsData, loadAssessment, and loadAssessmentFromTransfer) and updates the AssessmentCardSelectionStore state based on the payload. 

The `store-data-to-scan-node-result-converter` was updated to take in the `testKey` as a param, since we would want to populate the AssessmentCardSelectionStore with data from all the assessment tests, not just the selected test.

##### Motivation

Part of feature work

##### Context

It uses `store-data-to-scan-node-result-converter` to convert the assessment store data to card selection store data for the load actions.

Added `assessment-card-selection-store.ts` to the exclude list in `tsconfig.strictNullChecks.json` which is not ideal, but there were failing null checks from `initial-assessment-store-data-generator.ts` (already in the exclude list) that would require changing a lot of files that could potentially impact existing functionality.

#### Pull request checklist

- [ ] Addresses an existing issue: #0000
- [x] Ran `yarn fastpass`
- [x] Added/updated relevant unit test(s) (and ran `yarn test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] PR title *AND* final merge commit title both start with a semantic tag (`fix:`, `chore:`, `feat(feature-name):`, `refactor:`). See `CONTRIBUTING.md`.
- [n/a] (UI changes only) Added screenshots/GIFs to description above
- [n/a] (UI changes only) Verified usability with NVDA/JAWS

### CODE REVIEW

File: src/background/assessment-card-controller.ts
Comment (sfoslund) [inline src/background/assessment-card-controller.ts:19]: Left behind from debugging?

File: src/background/stores/assessment-card-selection-store.ts
Comment (sfoslund) [inline src/background/stores/assessment-card-selection-store.ts:68]: 2 things here: 
- The normal pattern would be to override the `generateDefaultState` method instead of initialize. What is the reason for using initialize here? It looks like we're pretty much just duplicating the persistent store initialize logic. 
- This doesn't include the logic to get the state based off of the assessment store state- are you going to add that in a separate PR?

File: src/background/stores/assessment-card-selection-store.ts
Comment (sfoslund) [inline src/background/stores/assessment-card-selection-store.ts:82]: nit: check that payload.assessmentStoreData.assessment is non-null and has more than 0 entries here too

File: src/background/assessment-card-controller.ts
Comment (sfoslund) [inline src/background/assessment-card-controller.ts:8]: It doesn't look like this is ever getting constructed/ initialized yet. Is there any reason to put this off to another PR? Have you tested end to end with this being initialized?

File: src/common/messages.ts
Comment (sfoslund) [inline src/common/messages.ts:241]: I believe we will want to have 2 assessment card controllers eventually- one for assessment data and one for quick assess data. As such, we will have to duplicate a lot of this code including the following:

- Add an additional AssessmentStoreChanged message to the QuickAssesCardSelection section
- Mirror the logic that this PR is adding in the assessment-card-selection-action-creator in the quick-assess-card-selection-action-creator
- Initialize the controller twice- once with the assessmentCardSelectionMessageCreator and once with the quickAssessCardSelectionMessageCreator

If you prefer, I'm okay with putting this off until another PR though. I forgot to make the message creator able to handle both assessment and quick assess messages in the last PR I made, so those changes are not yet in main for you to integrate with, which would be a blocker. Up to you if you prefer to wait until my next PR is merged or if you want to just port that change to this PR. Here's the necessary change: https://github.com/microsoft/accessibility-insights-web/pull/6501/files#diff-63c0ad049698d3b0edac710cab5bfa621ff664673990ab907f5280ca8d0bb5ceL11-L90

File: src/tests/unit/tests/background/stores/assessment-card-selection-store.test.ts
Comment (sfoslund) [inline src/tests/unit/tests/background/stores/assessment-card-selection-store.test.ts:144]: This test/ test name seems misleading- the title implies that we're setting the state based on the assessment data but we end up verifying that the final state is the expectedState, which is just the default state. Am I misunderstanding?

File: src/tests/unit/tests/background/stores/assessment-card-selection-store.test.ts
Comment (sfoslund) [inline src/tests/unit/tests/background/stores/assessment-card-selection-store.test.ts:184]: nit: Could probably do a it.each here with all of the cases where we expect the initial and expected state to end up the same- including when the assessmentStoreData.assessments are null and when they are empty.

**Review by sfoslund — COMMENTED** [review]
Logic looks good over all, a few small comments

File: src/background/stores/assessment-card-selection-store.ts
Comment (lisli1) [reply src/background/stores/assessment-card-selection-store.ts:82]: I was thinking about this too but wasn't completely sure if we wanted to clear the card selection store state if the assessment store state was cleared (on starting over assessment), but I'll double check how the assessment store state looks after such an event.

File: src/background/assessment-card-controller.ts
Comment (lisli1) [reply src/background/assessment-card-controller.ts:8]: I'll add this

File: src/tests/unit/tests/background/stores/assessment-card-selection-store.test.ts
Comment (sfoslund) [reply src/tests/unit/tests/background/stores/assessment-card-selection-store.test.ts:144]: Ah I see, thanks for the clarification!

File: src/background/stores/assessment-card-selection-store.ts
Comment (lisli1) [reply src/background/stores/assessment-card-selection-store.ts:82]: It appears that the assessment store state should never be empty or null, so I'll add the check for payload.assessmentStoreData.assessments as well. Thanks!

File: src/background/stores/assessment-card-selection-store.ts
Comment (lisli1) [reply src/background/stores/assessment-card-selection-store.ts:68]: Discussed offline; we decided that we can pass the persisted assessment store state in the constructor. Will update to make these changes

File: src/background/stores/assessment-card-selection-store.ts
Comment (waabid) [inline src/background/stores/assessment-card-selection-store.ts:83]: I am not sure if this is the right approach here due to, what I imagine, will be the perf implications; converting the assessment store completely after EACH change seems quite cost-intensive.

Was the approach of adding listeners to assessment actions (and updating accordingly) not feasible?

I suspect we'll have to do a conversion like this every so often (after say load assessment) but not necessarily every change.

File: src/background/stores/assessment-card-selection-store.ts
Comment (waabid) [inline src/background/stores/assessment-card-selection-store.ts:53]: nit: I think you could do this in `initialize` since that is usually where we load the persisted data.

File: src/background/stores/assessment-card-selection-store.ts
Comment (waabid) [inline src/background/stores/assessment-card-selection-store.ts:288]: Is there a reason we are validating so much of the payload? We don't usually do that ...

File: src/background/stores/assessment-card-selection-store.ts
Comment (waabid) [reply src/background/stores/assessment-card-selection-store.ts:269]: I think false is best for now; in Assessment we don't usually enable the visual helper by default (but do run a scan). I could see us changing that though, if PMs want that change.

File: src/background/stores/assessment-card-selection-store.ts
Comment (waabid) [inline src/background/stores/assessment-card-selection-store.ts:104]: nit: it probably makes sense to extract this to a private `getDefaultTestSelectionData` method.

File: src/common/store-data-to-scan-node-result-converter.ts
Comment (waabid) [inline src/common/store-data-to-scan-node-result-converter.ts:39]: I'm not sure I follow why this is necessary? This code is invoked in the fast pass (needs review/issues) card selection stores as well and would be hesitant to make any behavioral changes that might impact them.

File: src/tests/unit/tests/background/stores/assessment-card-selection-store.test.ts
Comment (waabid) [inline src/tests/unit/tests/background/stores/assessment-card-selection-store.test.ts:546]: nit: you should look to move these stub keys/ids out if you use them in multiple tests to reduce duplication and also makes it easier for future modifications.

File: src/common/store-data-to-scan-node-result-converter.ts
Comment (lisli1) [reply src/common/store-data-to-scan-node-result-converter.ts]: Removed

File: src/background/stores/assessment-card-selection-store.ts
Comment (waabid) [inline src/background/stores/assessment-card-selection-store.ts:302]: I'm okay with leaving this as is but I wonder if in the future we'll need to change this to 'go through each test key and just set it to default'. Regardless, I think `onScanCompleted` would be able to handle a completely empty object as well.

**Review by waabid — APPROVED** [review]

**Pull request merged**
